[PERF] Vectorize select_dtypes

datajanko opened the pull request:
- [x] closes #28317
- [x] tests added / passed (no new tests added, test_dtypes.py, passes locally) 
- [x] passes `black pandas` (passed locally)
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff` (passed locally)
- [x] whatsnew entry
- [x] ASV Benchmark
- [x] final refactoring

I was able to run this code for 10million columns in less than a second, which should solve the mentioned problem sufficiently well. 

<details><summary>ASV Benchmark:</summary>
```
[ 75.00%] ··· frame_methods.SelectDtypes.time_select_dtypes                                                                                                                                               ok
[ 75.00%] ··· ======== =============
                 n
              -------- -------------
                 10      1.53±0.2ms
                100     1.48±0.03ms
                1000    1.50±0.06ms
               10000    1.89±0.04ms
               100000    6.55±0.2ms
              ======== =============

[ 75.00%] · For pandas commit ad9fe5d7 <master^2> (round 2/2):
[ 75.00%] ·· Building for conda-py3.6-Cython-matplotlib-numexpr-numpy-odfpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt...
[ 75.00%] ·· Benchmarking conda-py3.6-Cython-matplotlib-numexpr-numpy-odfpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[100.00%] ··· frame_methods.SelectDtypes.time_select_dtypes                                                                                                                                               ok
[100.00%] ··· ======== ============
                 n
              -------- ------------
                 10     2.04±0.2ms
                100     10.0±0.3ms
                1000     91.2±2ms
               10000     898±50ms
               100000   8.93±0.08s
              ======== ============

       before           after         ratio
     [ad9fe5d7]       [ef1b9a49]
     <master^2>       <28317-select-dtypes-performance>
-      10.0±0.3ms      1.48±0.03ms     0.15  frame_methods.SelectDtypes.time_select_dtypes(100)
-        91.2±2ms      1.50±0.06ms     0.02  frame_methods.SelectDtypes.time_select_dtypes(1000)
-        898±50ms      1.89±0.04ms     0.00  frame_methods.SelectDtypes.time_select_dtypes(10000)
-      8.93±0.08s       6.55±0.2ms     0.00  frame_methods.SelectDtypes.time_select_dtypes(100000)
```
I hope this is sufficient
</details>

datajanko committed:
Merge pull request #1 from pandas-dev/master

updates pandas

datajanko committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas

datajanko committed:
more vectorization in select_dtypes

datajanko committed:
adds select_dtypes benchmark

datajanko committed:
parametrized benchmark

datajanko committed:
adds what's new entry

datajanko committed:
refactoring

datajanko commented:
A review would be nice, as well as maybe a better name for `filter_unique_dtypes_on`

datajanko commented:
Just the ASV benchmark with the most recent code versions
<detail>
```
· Creating environments
· Discovering benchmarks
·· Uninstalling from conda-py3.6-Cython-matplotlib-numexpr-numpy-odfpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt.
·· Building 16554c61 <28317-select-dtypes-performance> for conda-py3.6-Cython-matplotlib-numexpr-numpy-odfpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt............................................
·· Installing 16554c61 <28317-select-dtypes-performance> into conda-py3.6-Cython-matplotlib-numexpr-numpy-odfpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt..
· Running 2 total benchmarks (2 commits * 1 environments * 1 benchmarks)
[  0.00%] · For pandas commit ad9fe5d7 <master^2> (round 1/2):
[  0.00%] ·· Building for conda-py3.6-Cython-matplotlib-numexpr-numpy-odfpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt...
[  0.00%] ·· Benchmarking conda-py3.6-Cython-matplotlib-numexpr-numpy-odfpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[ 25.00%] ··· Running (frame_methods.SelectDtypes.time_select_dtypes--).
[ 25.00%] · For pandas commit 16554c61 <28317-select-dtypes-performance> (round 1/2):
[ 25.00%] ·· Building for conda-py3.6-Cython-matplotlib-numexpr-numpy-odfpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt...
[ 25.00%] ·· Benchmarking conda-py3.6-Cython-matplotlib-numexpr-numpy-odfpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[ 50.00%] ··· Running (frame_methods.SelectDtypes.time_select_dtypes--).
[ 50.00%] · For pandas commit 16554c61 <28317-select-dtypes-performance> (round 2/2):
[ 50.00%] ·· Benchmarking conda-py3.6-Cython-matplotlib-numexpr-numpy-odfpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[ 75.00%] ··· frame_methods.SelectDtypes.time_select_dtypes                                                                                                                                               ok
[ 75.00%] ··· ======== =============
                 n
              -------- -------------
                 10      1.37±0.1ms
                100     1.41±0.07ms
                1000     1.56±0.3ms
               10000     1.86±0.3ms
               100000    6.43±0.3ms
              ======== =============

[ 75.00%] · For pandas commit ad9fe5d7 <master^2> (round 2/2):
[ 75.00%] ·· Building for conda-py3.6-Cython-matplotlib-numexpr-numpy-odfpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt...
[ 75.00%] ·· Benchmarking conda-py3.6-Cython-matplotlib-numexpr-numpy-odfpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[100.00%] ··· frame_methods.SelectDtypes.time_select_dtypes                                                                                                                                               ok
[100.00%] ··· ======== ============
                 n
              -------- ------------
                 10     2.17±0.3ms
                100     10.3±0.2ms
                1000     95.5±4ms
               10000     919±30ms
               100000   9.46±0.03s
              ======== ============

       before           after         ratio
     [ad9fe5d7]       [16554c61]
     <master^2>       <28317-select-dtypes-performance>
-      2.17±0.3ms       1.37±0.1ms     0.63  frame_methods.SelectDtypes.time_select_dtypes(10)
-      10.3±0.2ms      1.41±0.07ms     0.14  frame_methods.SelectDtypes.time_select_dtypes(100)
-        95.5±4ms       1.56±0.3ms     0.02  frame_methods.SelectDtypes.time_select_dtypes(1000)
-        919±30ms       1.86±0.3ms     0.00  frame_methods.SelectDtypes.time_select_dtypes(10000)
-      9.46±0.03s       6.43±0.3ms     0.00  frame_methods.SelectDtypes.time_select_dtypes(100000)

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```
</detail>

jreback commented on pandas/core/frame.py:
```diff
@@ -3489,33 +3488,30 @@ def _get_info_slice(obj, indexer):
-                exclude_these.iloc[idx] = not any(map(f, exclude))
+        # We raise when both include and exclude are empty
+        # Hence, we can just shrink the columns we want to keep
+        keep_these = Series(True, index=self.columns)
```
this is significantly more complex and need refactoring

datajanko replied:
Could you elaborate a bit? Do you mean the comment? If so, I intended to keep it short as the code is, in my opinion more clearly in what it does. Previously we had the `exclude_these` array, but from the `and` concatenation at the we keep the columns in there. This didn't make too much sense. I would agree with the `filter_unique_dtypes_on` function though, but I don't see a good way to refactor it. Suggestions welcome.

WillAyd replied:
Instead of constructing a Series can you just use a lower level mask using an ndarray?

datajanko replied:
Will implement that. Actually, due to your emphasis on using a mask, I think we can also entirely get rid of  `_get_info_slice` some lines above. We don't need to create slices as we are using a mask.

datajanko replied:
I replaces the series by a boolean Numpy array

WillAyd commented on pandas/core/frame.py:
```diff
@@ -3489,33 +3488,30 @@ def _get_info_slice(obj, indexer):
+        keep_these = Series(True, index=self.columns)
+        unique_dtypes = self.dtypes.unique()
+
+        def filter_unique_dtypes_on(selection, unqiue_dtypes):
```
Can you annotate this function? Would help readability

datajanko replied:
you mean a small comment or a decorator (e.g. @staticmethod)

WillAyd replied:
Type hints if feasible

WillAyd commented on pandas/core/frame.py:
```diff
@@ -3489,33 +3488,30 @@ def _get_info_slice(obj, indexer):
+        keep_these = Series(True, index=self.columns)
+        unique_dtypes = self.dtypes.unique()
+
+        def filter_unique_dtypes_on(selection, unqiue_dtypes):
```
```suggestion
        def filter_unique_dtypes_on(selection, unique_dtypes):
```

small typo

WillAyd commented on doc/source/whatsnew/v0.25.2.rst:
```diff
@@ -6,6 +6,13 @@ What's new in 0.25.2 (October XX, 2019)
 These are the changes in pandas 0.25.2. See :ref:`release` for a full changelog
 including other versions of pandas.
 
+.. _whatsnew_0252.performance:
```
Can you move to v1.0.0?

WillAyd commented on asv_bench/benchmarks/frame_methods.py:
```diff
@@ -609,4 +609,15 @@ def time_dataframe_describe(self):
 
 
+class SelectDtypes:
+    params = [10, 100, 1000, 10000, 100000]
```
Can you post ASV results? Might not need all of these parameters, especially if they take significant time to run

datajanko replied:
check my comment above https://github.com/pandas-dev/pandas/pull/28447#issuecomment-531514169
If you have a suggestion on how to extract the results in a more readable/compact way, I'll change the comment

datajanko committed:
refactoring

adds Type Hints
removes now unnecessary code for selecting slices
uses iloc access
replaces list comprehension by loop

datajanko committed:
changes what's new to v 1.0.0 instead of 0.25.2

datajanko committed:
removes the smallest and the two biggest(slowest) benchmarks

datajanko committed:
replaces boolean series by numpy array

datajanko commented:
I'm not totally sure about the type annotations, specifically which type I should use for type. I think I tackled the other issues as well.

WillAyd commented on pandas/core/frame.py:
```diff
@@ -3489,33 +3479,34 @@ def _get_info_slice(obj, indexer):
+        keep_these = np.full(self.shape[1], True)
+
+        def extract_unique_dtypes_from_dtypes_list(
+            dtypes_list: List[Type], unique_dtypes: List[Type]
```
```suggestion
            dtypes_list: FrozenSet[Dtype], unique_dtypes: np.ndarray
```

Looking at code I don't think either of the arguments here are lists. You can import `Dtype` from pandas._typing and use that to subscript the first argument. The latter appears to be an ndarray, which unfortunately doesn't support subscripting

WillAyd commented on pandas/core/frame.py:
```diff
@@ -3489,33 +3479,34 @@ def _get_info_slice(obj, indexer):
+
+        def extract_unique_dtypes_from_dtypes_list(
+            dtypes_list: List[Type], unique_dtypes: List[Type]
+        ) -> List[Type]:
```
```suggestion
        ) -> List[Dtype]:
```

WillAyd commented on pandas/core/frame.py:
```diff
@@ -3489,33 +3479,34 @@ def _get_info_slice(obj, indexer):
+        # Hence, we can just shrink the columns we want to keep
+        keep_these = np.full(self.shape[1], True)
+
+        def extract_unique_dtypes_from_dtypes_list(
```
Hmm is this function even required? Would `np.isin` not offer similar functionality?

datajanko replied:
The DTypes are hierarchical, i.e. number contains e.g. timeldetas. In the tests we often see, `include=number` and `exclude=timedelta`. Somehow we need to take care of that.

datajanko committed:
fixes type annotations

TomAugspurger commented:
FYI, some CI checks are failing.

I think this is a generally promising approach. Will take a closer look later.

datajanko commented:
Sorry, I was at work. CI is only failing due to the type annotation. Will try to figure it put. The last commit has an unfortunate name. I just was fixing a wrong export from `pandas._typing`.

datajanko committed:
remove iteration as issubclass accepts tuples of classes

datajanko committed:
renaming

datajanko commented:
Hmm, sorry, I don't get the typing annotations here:

```
pandas/core/frame.py:3492: error: Argument 1 to "tuple" has incompatible type "FrozenSet[Union[str, Any, ExtensionDtype]]"; expected "Iterable[Union[type, Tuple[Any, ...]]]"
```

and tuple is called with
```
tuple(dtypes_set)
```
and from the type hints at the beginning:
```
dtypes_set: FrozenSet[Dtype]
```

Btw, I recharged to a list comprehension since I'm only having one loop now. But of course I would change back. Any hints on the typing related stuff?

WillAyd commented:
Seems weird - I would expect frozen set to be a subset of Iterable in the typing world. Any chance you see a bug report about it?

jreback commented on doc/source/whatsnew/v1.0.0.rst:
```diff
@@ -99,6 +99,7 @@ Performance improvements
 - Performance improvement in :func:`cut` when ``bins`` is an :class:`IntervalIndex` (:issue:`27668`)
 - Performance improvement in :meth:`DataFrame.corr` when ``method`` is ``"spearman"`` (:issue:`28139`)
 - Performance improvement in :meth:`DataFrame.replace` when provided a list of values to replace (:issue:`28099`)
+- Performance improvement in :meth:`DataFrame.select_dtypes` by using vectorization instead of iterating over a loop. With this improvement :meth:`DataFrame.select_dtypes` becomes feasible for dataframes with lots of columns. (:issue:`28317`)
```
can you remove the last sentence.

datajanko replied:
done

jreback reviewed (changes requested):
small comment, otherwise lgtm. ping on green.

datajanko committed:
make what'snew leaner

datajanko commented:
To file a bug report for the typing issue, not sure that this is a bug though, I'll try to create an easier example that shows the issue.

Right now I'm having:
```
from typing import FrozenSet, Iterable, Any, Union

from pandas._typing import ExtensionDtype, Dtype
import numpy as np

dtypes : FrozenSet[Dtype] = frozenset([np.numbers, np.float64])

tuple(dtypes)
``` 
which does not throw the Mypy error.

Or shall we just remove, the annotations and keep going?

WillAyd commented:
I think just add a `# type: ignore` to the line that is failing. Unfortunate this is failing but prefer to keep the annotations as they are at the very least good documentation

datajanko commented:
Okay, so I think I created and MVE:

```
from typing import List

l: List[int] = []
issubclass(object, tuple(l))
issubclass(object, l)
```

gives using Mypy
```
line 4: error: Argument 1 to "tuple" has incompatible type "List[int]"; expected "Iterable[Union[type, Tuple[Any, ...]]]"
line 5: error: Argument 2 to "issubclass" has incompatible type "List[int]"; expected "Union[type, Tuple[Union[type, Tuple[Any, ...]], ...]]"
```
The second line is only to show what is subclass expects. We have the analogous error message with Dtype. Can anybody explain this?  Is this a bug?

WillAyd commented:
@datajanko pretty confusing error message but aren't those invocations wrong? Should the second argument to `isinstance` be a type and not an instance thereof?

WillAyd commented:
Thinking through this I think the definition of `DType` in pandas._typing might be wrong and maybe that's actually throwing the error message. Right now it's `Union[str, np.dtype, "ExtensionDtype"]`, effectively allowing an instance of any of those types. But it might need to be `Type[Union[str, np.dtype, "ExtensionDtype"]]` to require the *class* instead of the instance (or something to that effect).

Tried that locally on your branch but that caused some more failures. For now I'd suggest just `type: ignoring" the failing line and we can open up a follow up issue to address the above

WillAyd commented:
cc @simonjayhawkins on previous comment

datajanko committed:
ignore type check

WillAyd commented on pandas/core/frame.py:
```diff
@@ -3489,33 +3479,35 @@ def _get_info_slice(obj, indexer):
+        # Hence, we can just shrink the columns we want to keep
+        keep_these = np.full(self.shape[1], True)
+
+        def extract_unique_dtypes_from_dtypes_set(
```
Hmm just thinking this over but can we just more succinctly write this as:

```
extracted_dtypes = list(filter(lambda x: issubclass(x.type, tuple(dtypes_set)), unique_dtypes))
```

? I think that one liner could replace a lot of the code here

datajanko replied:
I always thought that list comprehensions should be preferred in python. However, I just realized, that we could simply extract all dtypes directly and only once (and use intersection later). Choosing e..g `dtypes_set=include.union(exclude)`. But maybe this is what you meant.

WillAyd replied:
Yea. It just seems unnecessarily verbose the way it is now so I think can be condensed to help readability (though even as is its a good improvement over existing)

datajanko replied:
see comment below

WillAyd commented on pandas/core/frame.py:
```diff
@@ -3489,33 +3479,35 @@ def _get_info_slice(obj, indexer):
+        unique_dtypes = self.dtypes.unique()
+
+        if include:
+            included_dtypes = extract_unique_dtypes_from_dtypes_set(
```
If above is true would be nice to not duplicate this call in both branches

datajanko replied:
I'd assume that above is true: so shall I replace?

datajanko replied:
see comment below

WillAyd reviewed (commented):
Looks good. May be enough as is just curious if there's not a way to make more succinct

datajanko commented:
@WillAyd 

What works is the following:
```
        unique_dtypes = self.dtypes.unique()
        extracted_dtypes = lambda y:list(filter(lambda x: issubclass(x.type, tuple(y)), unique_dtypes))

        if include:
            keep_these &= self.dtypes.isin(extracted_dtypes(include))

        if exclude:
            keep_these &= ~self.dtypes.isin(extracted_dtypes(exclude))

        return self.iloc[:, keep_these.values]
```

note that extracted_dtypes is function. Do we prefer to `def` it?

WillAyd commented:
Nevermind then - I don't necessarily find that more readable

WillAyd reviewed (approved):
lgtm

datajanko commented:
@jreback ping since green

jreback commented:
thanks @datajanko nice patch!
